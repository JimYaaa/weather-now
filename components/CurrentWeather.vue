<script lang="ts" setup>
import type { CurrentWeather } from '~~/composables/useWeather'
import type { Gecoding } from '~~/composables/useLocation'

const props = defineProps({
    weather: {
        required: true,
        type: Object as PropType<CurrentWeather>
    },
    location: {
        required: true,
        type: Object as PropType<Gecoding | null> | null
    }
})
</script>

<template>
    <div>
        <div class="flex items-end">
            <div class="flex items-start color-white font-sans mr-10">
                <span class="text-20">{{ props.weather.temperature_celsius }}</span>
                <span class="text-10">{{ props.weather.temperature_celsius_unit }}</span>
            </div>

            <Icon class="color-white" :name="props.weather.weather_code" size="50" />
        </div>

        <div class="my-4">
            <p class="font-sans text-10 color-white">
                {{ location?.name }} 
                {{ location?.name !== location?.country ? `/ ${location?.country}` : '' }}
            </p>
        </div>

        <div class="flex justify-start items-center color-white">
            <div class="flex flex-col items-center mr-4">
               <Icon class="color-white mb-1" name="mdi:weather-windy" size="20" />
               <p>{{ props.weather.wind_speed + props.weather.wind_speed_unit }}</p>
            </div>

            <div class="flex flex-col items-center">
               <Icon class="color-white mb-1" name="mdi:water" size="20" />
               <p>{{ props.weather.relative_humidity }}</p>
            </div>
        </div>
    </div>
</template>