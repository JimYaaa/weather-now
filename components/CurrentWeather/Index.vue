<script lang="ts" setup>
import type { Gecoding } from '~~/entities/gecoding'
import type { CurrentWeather } from '~~/entities/weather'

const props = defineProps({
    weather: {
        required: true,
        type: Object as PropType<CurrentWeather>
    },

    location: {
        required: true,
        type: Object as PropType<Gecoding | null> | null
    },

    temperatureUnit: {
        required: true,
        type: String,
    }
})

const temperature = computed(() => {
    if (props.temperatureUnit === 'celsius') {
        return {
            value: props.weather.temperature_celsius,
            unit: props.weather.temperature_celsius_unit,
        }
    }

    return {
        value: props.weather.temperature_fahrenheit,
        unit: props.weather.temperature_fahrenheit_unit,
    }
})

defineExpose({ temperature })
</script>

<template>
    <div class="w-full flex justify-between items-start">
        <div>
            <div class="flex items-end">
                <div :key="props.temperatureUnit" class="flex items-start color-white font-sans mr-10">
                    <span class="text-16 md-text-20">{{ temperature.value }}</span>
                    <span class="text-8 md-text-10">{{ temperature.unit }}</span>
                </div>
            </div>
    
            <div class="my-6">
                <p class="font-sans text-4 md-text-6 color-white">
                    {{ location?.name }} 
                    {{ location?.name !== location?.country ? `/ ${location?.country}` : '' }}
                </p>
            </div>
    
            <div class="flex justify-start items-center color-white">
                <div class="flex flex-col items-center">
                   <Icon class="color-white mb-1 text-4 md-text-6" name="mdi:weather-windy" />
                   <p class="text-4 md-text-6 mt-2">{{ props.weather.wind_speed + props.weather.wind_speed_unit }}</p>
                </div>

                <div class="w-1px h-40px mx-4 bg-white"></div>
    
                <div class="flex flex-col items-center">
                   <Icon class="color-white mb-1 text-4 md-text-6" name="mdi:water" />
                   <p class="text-4 md-text-6 mt-2">{{ props.weather.relative_humidity }}</p>
                </div>
            </div>
        </div>

        <div>
            <Icon class="color-white text-30" :name="props.weather.weather_code" />
        </div>
    </div>
</template>
