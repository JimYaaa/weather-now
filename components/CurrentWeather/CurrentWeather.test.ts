import { mountSuspended } from '@nuxt/test-utils/runtime'
import { describe, it, expect } from 'vitest'
import CurrentWeather from './Index.vue'

describe('CurrentWeather', () => {
    const propsConfig = {
        weather: {
            weather_code: 'wi:day-sunny',
            temperature_celsius: 20,
            temperature_celsius_unit: '째C',
            temperature_fahrenheit: 50,
            temperature_fahrenheit_unit: '째F',
            relative_humidity: '100%',
            wind_speed: 100,
            wind_speed_unit: 'km/h',
        },
        location: {
            id: 123,
            name: 'Taiwan',
            latitude: 12,
            longitude: 34,
            country: 'Taiwan',
        },
    }

    it('when temperature unit is celsius', async () => {
        const component = await mountSuspended(CurrentWeather, {
            props: {
                weather: propsConfig.weather,
                location: propsConfig.location,
                temperatureUnit: 'celsius'
            },
        })

        const { weather } = propsConfig

        expect(component.html()).toContain('째C')
        expect(component.vm.temperature).toMatchObject({
            value: weather.temperature_celsius,
            unit: weather.temperature_celsius_unit
        })
        expect(component.html()).toMatchSnapshot()
    })

    it('when temperature unit is fahrenheit', async () => {
        const component = await mountSuspended(CurrentWeather, {
            props: {
                weather: propsConfig.weather,
                location: propsConfig.location,
                temperatureUnit: 'fahrenheit'
            }
        })

        const { weather } = propsConfig

        expect(component.html()).toContain('째F')
        expect(component.vm.temperature).toMatchObject({
            value: weather.temperature_fahrenheit,
            unit: weather.temperature_fahrenheit_unit
        })
        expect(component.html()).toMatchSnapshot()
    })
})