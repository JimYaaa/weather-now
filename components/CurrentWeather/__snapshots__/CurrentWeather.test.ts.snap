// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CurrentWeather > when temperature unit is celsius 1`] = `
"<div class="w-full flex justify-between items-start">
  <div>
    <div class="flex items-end">
      <div class="flex items-start color-white font-sans mr-10"><span class="text-16 md-text-20">20</span><span class="text-8 md-text-10">Â°C</span></div>
    </div>
    <div class="my-6">
      <p class="font-sans text-4 md-text-6 color-white">Taiwan </p>
    </div>
    <div class="flex justify-start items-center color-white">
      <div class="flex flex-col items-center"><span class="iconify i-mdi:weather-windy color-white mb-1 text-4 md-text-6" aria-hidden="true"></span>
        <p class="text-4 md-text-6 mt-2">100km/h</p>
      </div>
      <div class="w-1px h-40px mx-4 bg-white"></div>
      <div class="flex flex-col items-center"><span class="iconify i-mdi:water color-white mb-1 text-4 md-text-6" aria-hidden="true"></span>
        <p class="text-4 md-text-6 mt-2">100%</p>
      </div>
    </div>
  </div>
  <div><span class="iconify i-wi:day-sunny color-white text-30" aria-hidden="true"></span></div>
</div>"
`;

exports[`CurrentWeather > when temperature unit is fahrenheit 1`] = `
"<div class="w-full flex justify-between items-start">
  <div>
    <div class="flex items-end">
      <div class="flex items-start color-white font-sans mr-10"><span class="text-16 md-text-20">50</span><span class="text-8 md-text-10">Â°F</span></div>
    </div>
    <div class="my-6">
      <p class="font-sans text-4 md-text-6 color-white">Taiwan </p>
    </div>
    <div class="flex justify-start items-center color-white">
      <div class="flex flex-col items-center"><span class="iconify i-mdi:weather-windy color-white mb-1 text-4 md-text-6" aria-hidden="true"></span>
        <p class="text-4 md-text-6 mt-2">100km/h</p>
      </div>
      <div class="w-1px h-40px mx-4 bg-white"></div>
      <div class="flex flex-col items-center"><span class="iconify i-mdi:water color-white mb-1 text-4 md-text-6" aria-hidden="true"></span>
        <p class="text-4 md-text-6 mt-2">100%</p>
      </div>
    </div>
  </div>
  <div><span class="iconify i-wi:day-sunny color-white text-30" aria-hidden="true"></span></div>
</div>"
`;
